File: trie/trie.cpp
#include <boost/shared_ptr.hpp>
#include <iostream>
#include <fstream>
#include <string>
#include <assert.h>
#include <vector>

using namespace std;

class TrieNode;
typedef boost::shared_ptr<TrieNode> TrieNodePtr;

class TrieNode
{
    public:
        TrieNode(char _c = 0): c(_c), is_word(false) {}
        inline TrieNodePtr child(char _c) { return children[_c]; }
        inline void new_child(char _c) { children[_c] = TrieNodePtr(new TrieNode(_c)); }

        char c;
        bool is_word;
        TrieNodePtr children[128];

};

//-----------------------------------------------------------------------------

class Trie
{
    public:
        Trie(): root(new TrieNode()) {}
        void load(const string& file);
        void add(const string& word);
        bool search(const string& word) const;
        void suffix(TrieNodePtr n, string& s, vector<string>& v) const;
        vector<string> similar(const string& prefix) const;

    private:
        TrieNodePtr root;
};

void Trie::load(const string& file) {
    ifstream fs;
    fs.open(file.c_str(), ifstream::in);

    string word;
    while (getline(fs, word)) {
        add(word);
    }
    fs.close();
}

void Trie::add(const string& word) {
    TrieNodePtr n = root;

    for (int i = 0; i < word.length(); ++i) {
        char c = word[i];
        if (!n->child(c)) {
            n->new_child(c);
        }
        n = n->child(c);
    }
    n->is_word = true;
}

bool Trie::search(const string& word) const {
    TrieNodePtr n = root;
    int l = word.length();

    for (int i = 0; i < l; ++i) {
        char c = word[i];
        if (!n->child(c))
            return false;

        n = n->child(c);
    }
    return n->is_word;
}

void Trie::suffix(TrieNodePtr n, string& s, vector<string>& v) const {
    if (n) {
        s.push_back(n->c);

        // Found a word
        if (n->is_word) {
            v.push_back(s);
        }

        for (char c = 0; c < (char)127; ++c) {
            if (n->child(c)) {
                suffix(n->child(c), s, v);
                // Backtrack
                s.erase(s.length()-1, 1);
            }
        }
    }
}

vector<string> Trie::similar(const string& prefix) const {
    vector<string> v;
    TrieNodePtr n = root;
    for (int i = 0; i < prefix.length(); ++i) {
        char c = prefix[i];
        if (!n->child(c))
            return v;

        n = n->child(c);

        // The prefix itself is a valid word
        if (i == prefix.length()-1 && n->is_word)
            v.push_back(prefix);
    }
    for (char c = 0; c < (char)127; ++c) {
        if (n->child(c)) {
            string s(prefix);
            suffix(n->child(c), s, v);
        }
    }
    return v;
}

//-----------------------------------------------------------------------------

int main(int argc, char* argv[])
{
    string prefix("prep");
    if (argc > 1) {
        prefix = argv[1];
    }

    Trie t;
    t.load("enable1.txt");
    assert(t.search("dog") == true);
    assert(t.search("cog") == true);
    assert(t.search("fool") == true);
    assert(t.search("sage") == true);
    assert(t.search("khoi") == false);
    assert(t.search("zymogen") == true);
    assert(t.search("nguyen") == false);
    cout << "Success!" << endl;

    vector<string> v = t.similar(prefix);
    for (int i = 0; i < v.size(); ++i) {
        cout << v[i] << endl;
    }

    return 0;
}

File: trie/Trie.java
import org.apache.commons.lang3.StringUtils;
import org.apache.commons.lang3.text.StrBuilder;

import java.io.BufferedReader;
import java.io.FileReader;
import java.io.IOException;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

/**
 * Trie structure for string searches
 */
public class Trie {

    /**
     * Nodes in a trie structure
     */
    private class TrieNode {
        private char c;
        private Map<Character, TrieNode> children = new HashMap<>();
        private boolean isWord = false;

        private TrieNode() {
        }

        private TrieNode(char c) {
            this.c = c;
        }

        private TrieNode child(char c) {
            return children.get(c);
        }

        private void addChild(char c) {
            children.put(c, new TrieNode(c));
        }
    }

    // TODO: command line arg?
    private static final String DICTIONARY_FILE = "enable1.txt";

    // Root of the entire trie structure. Never null.
    private TrieNode root = new TrieNode();


    /**
     * Load dictionary of English words from a file
     *
     * @param file containing words to load
     * @throws IOException
     */
    public void load(final String file) throws IOException {
        final BufferedReader br = new BufferedReader(new FileReader(file));
        String line;
        while ((line = br.readLine()) != null) {
            add(StringUtils.trim(line));
        }
    }

    /**
     * Checks for existence of a valid word in the trie
     *
     * @param word to look for
     * @return
     */
    public boolean contains(final String word) {
        final TrieNode node = get(word);
        return (node != null) && node.isWord;
    }

    /**
     * Find all words that share the same prefix
     *
     * @param prefix
     * @return
     */
    public List<String> similar(final String prefix) {
        final List<String> list = new ArrayList<>();

        // Get all words that share the same prefix
        suffix(get(prefix), prefix, list);
        return list;
    }

    /**
     * Given a node, find all words that share the node with a common prefix
     *
     * @param node starting node to start finding suffixes
     * @param prefix to search words with
     * @param list to hold words found
     */
    private void suffix(final TrieNode node, final String prefix, final List<String> list) {
        if (node != null) {
            if (node.isWord) {
                list.add(prefix);
            }
            for (final TrieNode childNode : node.children.values()) {
                suffix(childNode, prefix + childNode.c, list);
            }
        }
    }

    /**
     * Add a word to the trie
     *
     * @param word to add
     */
    private void add(final String word) {
        if (StringUtils.isBlank(word)) {
            return;
        }

        TrieNode node = root;
        for (int i = 0; i < word.length(); ++i) {
            final char c = word.charAt(i);
            if (node.child(c) == null) {
                node.addChild(c);
            }
            node = node.child(c);
        }
        node.isWord = true;
    }

    /**
     * Get the furthest descendant of root node in the trie that composes the word. In essence, the last node
     * in the sequence of trie nodes that make up the provided word.
     *
     * @param word
     * @return
     */
    private TrieNode get(final String word) {
        if (StringUtils.isBlank(word)) {
            return null;
        }
        TrieNode node = root;
        for (int i = 0; i < word.length(); ++i) {
            node = node.child(word.charAt(i));
            if (node == null) {
                return null;
            }
        }
        return node;
    }

    public static void main(final String[] args) throws Exception {
        if (args.length < 1) {
            System.out.println("Prefix string must be provided");
            return;
        }

        final Trie trie = new Trie();
        trie.load(Trie.DICTIONARY_FILE);
        assert(trie.contains("grape"));
        assert(trie.contains("lion"));
        assert(!trie.contains("khoi"));
        assert(trie.contains("stop"));

        System.out.println(new StrBuilder().appendWithSeparators(trie.similar(args[0]), ", "));
    }
}
